--- a/chat.jinja
+++ b/chat.jinja
@@ -90,7 +90,7 @@
     {%- set content = render_content(message.content, true) | trim %}
     {%- if loop.index0 < leading.count %}
     {%- elif message.role == 'system' or message.role == 'developer' %}
-        {{- raise_exception('System/developer message must be at the beginning.') }}
+        {{- '<|im_start|>system\n' + content + '<|im_end|>\n' }}
     {%- elif message.role == 'user' %}
         {{- '<|im_start|>' + message.role + '\n' + content + '<|im_end|>' + '\n' }}
     {%- elif message.role == "assistant" %}
