--- a/chat.jinja
+++ b/chat.jinja
@@ -136,10 +136,11 @@
                 {%- else %}
                     {{- '\n<tool_call>\n<function=' + tool_call.name + '>\n' }}
                 {%- endif %}
-                {%- if tool_call.arguments is defined and tool_call.arguments != '' %}
-                    {%- for args_name, args_value in tool_call.arguments|items %}
+                {%- if tool_call.arguments is defined and tool_call.arguments is mapping %}
+                    {%- for args_name in tool_call.arguments %}
+                        {%- set args_value = tool_call.arguments[args_name] %}
                         {{- '<parameter=' + args_name + '>\n' }}
-                        {%- set args_value = args_value | string if args_value is string else args_value | tojson | safe %}
+                        {%- set args_value = args_value | tojson | safe if args_value is mapping or (args_value is sequence and args_value is not string) else args_value | string %}
                         {{- args_value }}
                         {{- '\n</parameter>\n' }}
                     {%- endfor %}
